### DISCUSSION TITLE
Using or_ in relationship's join

### DISCUSSION THREAD

**Discussion by Makistos** [sqlalchemy/sqlalchemy#9815] [Usage Questions] [answered]
Hi,

I have a table structure basically looking like this:
There is a Part table with foreign keys to Work and Edition and a Contributor table with a foreign key to Part. For this question it is also relevant that Part has a foreign key to ShortStory and Contributor has a foreign key to Role.

Now I would like a Contributions relationship to Work and Edition with the following conditions: firstly, shortstory_id needs to be null and secondly, for Work, role_id needs to be either 1 or 3 and and for edition any other value but 1 and 3.

For Work I can do this with either role_id 1 or role_id 3 like this:
```
    contributions = relationship('Contributor', secondary='part',
                                 primaryjoin='and_(Part.work_id == Work.id, Part.shortstory_id == None, Contributor.role_id == 1)',
                                 viewonly=True, uselist=True)
```
but I can't figure out how to add role_id == 3 to this. Anyone know how to do this? Edition is easy as it only needs and_.

**Comment by zzzeek** [comment, marked as answer]
you can embed or_() inside and_() like this:

     primaryjoin='and_(Part.work_id == Work.id, Part.shortstory_id == None, or_(Contributor.role_id == 1, Contributor.role_id ==3))'

?

you could also use `in_()`:

     primaryjoin='and_(Part.work_id == Work.id, Part.shortstory_id == None, Contributor.role_id.in_([1, 3]))'

**Comment by Makistos** [comment]
That seemed to do the trick! Quick followup, can I select just distinct values as well somehow? I tried distinct_target_key=True but that didn't seem to do the trick.

**Reply by zzzeek** [reply]
relationships are loading object-rows that have primary key columns, and the Contributor objects themselves would be in the identity map on primary key.   so if you are seeing the same Contributor object more than once in the collection, I would use [collection_class=set](https://docs.sqlalchemy.org/en/20/orm/collection_api.html#custom-collections)  and be done with it.

**Comment by Makistos** [comment]
That doesn't seem to work, contents still look like:
```
0:
{'real_person': None, 'person': {'name': 'Asimov, Isaac', 'id': 2687.0}, 'description': None, 'role': {'name': 'Kirjoittaja', 'id': 1}}
1:
{'real_person': None, 'person': {'name': 'Asimov, Isaac', 'id': 2687.0}, 'description': None, 'role': {'name': 'Kirjoittaja', 'id': 1}}
2:
{'real_person': None, 'person': {'name': 'Asimov, Isaac', 'id': 2687.0}, 'description': None, 'role': {'name': 'Kirjoittaja', 'id': 1}}
```
P.S. I just realised I use a rather old version of SQLAlchemy (1.4.5). 

Ah, I know why this is. That print is the data after it has gone through Marshmallow and the model doesn't include it but there's also a part_id which is unique -> these contributions are different.

**Reply by zzzeek** [reply]
this doesnt have context for me to really follow what you're doing

**Reply by Makistos** [reply]
Yes, sorry. I realise that. I made a mistake when I thought those rows all have the same contents. So distinct wouldn't have worked either.

**Reply by Makistos** [reply]
I'll just remove the (for my case) distinct values after the query in code.

**Reply by Makistos** [reply]
And just to explain so you are not left wondering:

Each work can have multiple editions. I originally thought back when I wrote the first version that there could be different authors in some cases for different editions so there is a part for every work - edition combination and a contributor for each part. So even though authors now can only be attached to works and not editions there are still multiple rows with that author info. For the front-end I don't want to show one author row for every edition, just the one. I then save this info at the backend for every edition again.
